gpt4all-chat/src/database.cpp

```diff
@@ -1937,20 +2078,222 @@ QList<int> Database::searchEmbeddings(const std::vector<float> &query, const QLi
     return chunkIds;
 }
 
-void Database::retrieveFromDB(const QList<QString> &collections, const QString &text, int retrievalSize,
-    QList<ResultInfo> *results)
+QList<int> Database::searchEmbeddings(const std::vector<float> &query, const QList<QString> &collections,
+    int nNeighbors)
 {
-#if defined(DEBUG)
-    qDebug() << "retrieveFromDB" << collections << text << retrievalSize;
+    QSqlQuery q(m_db);
+    if (!q.exec(GET_COLLECTION_EMBEDDINGS_SQL.arg(collections.join("', '")))) {
+        qWarning() << "Database ERROR: Failed to exec embeddings query:" << q.lastError();
+        return {};
+    }
+    return searchEmbeddingsHelper(query, q, nNeighbors);
+}
+
+QList<int> Database::scoreChunks(const std::vector<float> &query, const QList<int> &chunks)
+{
+    QList<QString> chunkStrings;
+    for (int id : chunks)
+        chunkStrings << QString::number(id);
+    QSqlQuery q(m_db);
+    if (!q.exec(GET_CHUNK_EMBEDDINGS_SQL.arg(chunkStrings.join("', '")))) {
+        qWarning() << "Database ERROR: Failed to exec embeddings query:" << q.lastError();
+        return {};
+    }
+    return searchEmbeddingsHelper(query, q, chunks.size());
+}
+
+QList<Database::BM25Query> Database::queriesForFTS5(const QString &input)
+{
+    // Escape double quotes by adding a second double quote
+    QString escapedInput = input;
+    escapedInput.replace("\"", "\"\"");
+
+    static QRegularExpression spaces("\\s+");
+    QStringList oWords = escapedInput.split(spaces, Qt::SkipEmptyParts);
+
+    QList<BM25Query> queries;
+
+    // Start by trying to match the entire input
+    BM25Query e;
+    e.type = 0; // exact
```

Comment:
Please either use a boolean (`bool isExact`) or an enumeration, so the type field can be self-documenting.